File: binary-tree/bintree.cpp
#include <iostream>
#include <stdlib.h>
using namespace std;

typedef struct tnode* Bintree;
typedef struct tnode{
    int key;
    Bintree left;
    Bintree right;
} Node;

#define left(node)  (node)->left
#define right(node) (node)->right
#define root(node)  (node)->key
#define empty(node) (node==NULL)

Bintree alloc(int x){
    Bintree new_node = (Bintree)malloc(sizeof(Node));
    if(empty(new_node)) return NULL;
    root(new_node) = x;
    left(new_node) = NULL;
    right(new_node) = NULL;
    return new_node;
} 

void makeBintree(Bintree *T){
    *T = NULL;
}

void preorder(Bintree t){
    if(empty(t)) return;
    preorder(left(t));
    cout<<root(t)<<" ";
    preorder(right(t));
}

void insert(Bintree *node,int x){
    if(empty(*node)) {
        *node=alloc(x);
        return;
    }
    if(root(*node)==x) {
        return;
    }
    
    if(x<root(*node)) insert(&left(*node),x);
    else insert(&right(*node),x);

}

Bintree search(Bintree T,int x){
    
    if(!empty(T)){
        if(x==root(T)) return T;
        else if(x<root(T)) return search(left(T),x);
        else return search(right(T),x);
    }
     return NULL;
    
}

int height(Bintree T){
    if(empty(T)) return 0;
    else return max((1+height(left(T))),(1+height(right(T))));
}

int totalNode(Bintree T){
    if(empty(T)) return 0;
    else return 1+totalNode(left(T)) + totalNode(right(T));
}

void addLeafMost(Bintree *T, Bintree *L){
    if(empty(left(*T))) left(*T) = *L;
    else addLeafMost(&left(*T),L);
}

void delNode(Bintree *T,int x){
    if(empty(*T)) return;
    
    if(x==root(*T)){
        if(empty(left(*T)) && empty(right(*T))) {
            *T = NULL;
            return;
        }

        Bintree temp = *T;
        *T = right(*T);
        addLeafMost(T,&left(temp));
        temp = NULL;
    
    }
    else if(x>root(*T)) delNode(&right(*T),x);
    else delNode(&left(*T),x);
    
}

int main(){
    Bintree root;
    makeBintree(&root);
    insert(&root,50);
    insert(&root, 30); 
    insert(&root, 20); 
    insert(&root,50);
    insert(&root, 40); 
    insert(&root, 70); 
    insert(&root, 60); 
    insert(&root, 80);
    // cout<<totalNode(root);
    // cout<<height(root);
    // cout<<search(root,60);
    // preorder(root);
    // delNode(&root,30);
    // cout<<"\n";
    preorder(root);
    return 0;
}

File: binary-tree/bintree.py
class Tree:
    def __init__(self,key):
        self.root = key
        self.left = None
        self.right = None
    
    def add(self,key):
        if self is None:
            return Tree(key)
        else:
            if(key<self.root):
                self.left = Tree.add(self.left,key)
            elif (key>self.root):
                self.right = Tree.add(self.right,key)
        return self

    def preorder(self):
        if self:
            print(self.root,end=" ")
            Tree.preorder(self.left)
            Tree.preorder(self.right)

    def draw(self,i=1):
        if self:
            print("|_"*i,"{}   lev.{}".format(self.root,i-1))
            Tree.draw(self.left,i+1)
            Tree.draw(self.right,i+1)

    def search(self,key):
        if self:
            if self.root == key:
                return self
            elif key<self.root :
                return Tree.search(self.left,key)
            elif key>self.root :
                return Tree.search(self.right,key)
        else:
            return False

    def __str__(self):
        return "\nroot : {root} \nleft : {left}  \nright: {right}".format(
            root=self.root,left=self.left,right=self.right
        )

if __name__ == "__main__":
    root = Tree(50)
    root.add(50)
    root.add(30) 
    root.add(20) 
    root.add(50)
    root.add(40) 
    root.add(70) 
    root.add(60) 
    root.add(80)
    # root.draw()
    print(root)